keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,546 @@
+kube-scheduler, the controller can make a decision based on the stopped Job
+queue (including Job preemption). Each Job could request a different amount of
+resources, so the higher-level controller may also want to resize the cluster
+to just fit the Job it's going to run. Regardless of what logic the controller
+uses to queue jobs, being able to stop Jobs indefinitely and then resume them
+later is important.
+
+### Notes/Constraints/Caveats
+
+System administrators may want to restrict who can set/modify the `stopped`
+field when creating or updating Jobs. This can be achieved with validating
+and/or mutating webhooks.
+
+When a Job is stopped, the time is recorded as part of the Job status.  Users
+can infer how long a Job has been paused for through this field. This can be
+useful when making decisions around which Job should be resumed.
+
+A Job that is complete cannot be stopped.
+
+### Risks and Mitigations
+
+1. Stopping an active Job deletes *all* active pods. Users must design their
+   application to gracefully handle this.
+
+## Design Details
+
+We propose adding a `stopped` field to the JobSpec API:
+
+```go
+type JobSpec struct {
+	// Stopped specifies whether the Job controller should create Pods or not. If
+	// a Job is created in the stopped state, no Pods are created by the Job
+	// controller. If a Job enters the stopped state (i.e. the flag goes from
+	// false to true), the Job controller will delete all active Pods associated
+	// with this Job. Users must design their workload to gracefully handle this.
+	// This is an alpha field and requires enabling the StoppedJob feature gate.
+	// Defaults to false.
+	Stopped *bool `json:"stopped,omitempty"`
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager abstains from creating Pods even if the Job's parallelism
+requirements aren't met. If the Job is already active and is updated with
+`stopped: true`, the Job controller calls Delete on all its active Pods. This
+causes the kubelet to send a SIGTERM signal and completely remove the Pod after
+its graceful termination period is honoured. Pods terminated this way are
+considered a failure and the controller does not count terminated Pods towards
+completions. This behaviour is similar to decreasing parallelism to zero.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition type called
+"Stopped" as a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
```

Comment:
A generic condition should be used instead:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1353

Reply:
This struct already exists, we are not adding it.